### DISCUSSION TITLE
Sometimes filters just stop working

### DISCUSSION THREAD

**Discussion by TonyGravagno** [RealRaven2000/FiltaQuilla#281] [General]
Some of us encounter this situation a lot. Filters just stop working for a profile or folder.  

Let's use this thread to try to understand and remedy the issue for ourselves and all Tb users.  
Please add comments as you get solid information. Please don't add a "me too" ... we know it's a general issue ... unless you know that the issue happens after some specific event.

Why is this here? Because this repo is about filtering, and we're all motivated to have our filters work as we expect, whether they are processed with the addons here or not. It's possible that we might find that some flag needs to be reset before all filters are run, and that might be possible in one of Axel's addons or in our own filters with JavaScript.

Axel https://github.com/RealRaven2000/FiltaQuilla/issues/260#issuecomment-2092691052

> ... it turned out that just this single folder was broken, the filter worked on other folders with IMAP; not quite sure what broke it thoug and how to repair that. On other profiles I was able to read body without any problems - go figure.

> I think one thing we do have is that Thunderbird profiles or even just single folders may "break" (at least from the perspective of filters) from time to time which is really not a good look for Thunderbird. 

> They do work on replacing the GLODA (global data store for all mails) completely but this is going to take a while because they have 20 year old architectural "debt" to Netscape, and they are NOT going to monkey patch but rewrite it from the ground up.

The workaround for now is simply to restart Thunderbird. Yeah, "not a good look".

What I've seen: One or more filters can process a mail item, get a successful match which stops further match testing, but then the actions aren't processed, except for Stop Filter Execution. "Train as ..." and "Move Message" are not processed.

@RealRaven2000 - is your reference to the GLODA a guess at where the problem might be, or is that a/the confirmed source of the problem?

If the source is still unknown, perhaps we can help to figure out a pattern:

1. Does it happen only after modifying filters, or can it happen purely when receiving with no config changes?
1. Does it happen only on a Manual filter run, or does it happen during receive too?
1. Does it break only after one of the following or after some other event?
    - menu > Tools > Run Filters on Folder?
    - topmenubar > quickFilters > Run Filters?
    - menu > Tools > Message Filters > Run Now?
1. Is this in any way related to addons, especially @RealRaven2000 addons? Or can his happen in Tb without addons?
1. Might everything break if there is a Move Message and the folder no longer exists?
1. Does this only happen when Move Message is the first action, or second?
1. Does it still happen if all Move Message actions are changed to Move Later?
1. Might this happen if messages are inbound from the server at the same time that other messages are being manually filtered?
1. What other pattern might be valid?

We might not find a pattern because it might happen only after something random like a specific dialog is opened, or menu items aren't used yet, or a second tab needs to be opened and closed. But try to be aware of such things.

We might not find "the" issue, but if we can identify a single trigger, that information might lead to understanding other factors that lead to the same result.

Is there already a Thunderbird bug ticket on this?

Is there a procedure to try to trace the issue?

Is there anything other than a restart that we can do to get around this?

On the down side, I believe the Thunderbird team is focused on final cleanup for v128, and even if we identify the source(s) of this problem, it might not be fixed for a long time to come. As Axel said, it might be related to the GLODA.

Thanks and Good Luck!

**Comment by RealRaven2000** [comment]
> 
> Axel [#260 (comment)](https://github.com/RealRaven2000/FiltaQuilla/issues/260#issuecomment-2092691052)
> 
> > ... it turned out that just this single folder was broken, the filter worked on other folders with IMAP; not quite sure what broke it thoug and how to repair that. On other profiles I was able to read body without any problems - go figure.
> 
> > I think one thing we do have is that Thunderbird profiles or even just single folders may "break" (at least from the perspective of filters) from time to time which is really not a good look for Thunderbird.
> 
> > They do work on replacing the GLODA (global data store for all mails) completely but this is going to take a while because they have 20 year old architectural "debt" to Netscape, and they are NOT going to monkey patch but rewrite it from the ground up.
> 
> The workaround for now is simply to restart Thunderbird. Yeah, "not a good look".
> 
> What I've seen: One or more filters can process a mail item, get a successful match which stops further match testing, but then the actions aren't processed, except for Stop Filter Execution. "Train as ..." and "Move Message" are not processed.
> 
> @RealRaven2000 - is your reference to the GLODA a guess at where the problem might be, or is that a/the confirmed source of the problem?

it's a guess. I a, not even sure whethe GLODA is involved during filtering. the grunt work of Filtering is done on the C++ layer, which needs to be debugged with a separate debugger (apart from the javascript browser debugger and the separate addon debuggers) it's a little above my paygrade as I haven't compiled Thunderbird in years and the code is DEEP. I really would leave this to Thunderbird Core developers if possible, if we can even find somebody who is specialized within this area.

> 
> If the source is still unknown, perhaps we can help to figure out a pattern:
> 
>     1. Does it happen only after modifying filters, or can it happen purely when receiving with no config changes?

I guess it can be triggered by receiving, but I am not 100% sure.

>     2. Does it happen only on a Manual filter run, or does it happen during receive too?

It may be both, and it may be that the success rate is higher when running manually. If it can be reliably reproduced by running Manually at least it would generate a path for debugging the problem buy sharing the profile with the Core team. (I think there are ways to report and reproduce this, while anonymizing data, but I am by no means an expert)

> 
>     3. Does it break only after one of the following or after some other event?
>        
>        * menu > Tools > Run Filters on Folder?
>        * topmenubar > quickFilters > Run Filters?
>        * menu > Tools > Message Filters > Run Now?

I think all these trigger the same function `goDoCommand("cmd_applyFilters")` which is a function of Thunderbird Core command handler:

my code from quickFilters.js:

```js
    switch (el.id) {
      case "quickfilters-toolbar-runbutton": // fall-throughs
      case "quickfilters-runFilters":
      case "quickfilters-current-runbutton":
      case RUNFILTERFROMTREE_ID:
          window.quickFilters.onApplyFilters();
        break;
```

```js
  onApplyFilters: function onApplyFilters(silent) {

    // does this work in non-inbox current folder?
    // Get the folder where filters should be defined, if that server
    // can accept filters.
    const util = quickFilters.Util,
          Ci = Components.interfaces,
          Cc = Components.classes,
          filterService = MailServices.filters; // nsIMsgFilterService

          
    let folder = util.getCurrentFolder(),
        msgWindow = Cc["@mozilla.org/messenger/msgwindow;1"].createInstance(Ci.nsIMsgWindow);
        
    //.....   

    if (folder.flags & util.FolderFlags.Inbox) {
      // MsgApplyFilters();
      goDoCommand("cmd_applyFilters");
    }
```

> 
>     4. Is this in any way related to addons, especially @RealRaven2000 addons? Or can his happen in Tb without addons?

If it uses standard search terms and actions (so none of the actions / search terms provided by  FiltaQuilla) then yes, this will absolutely also be happening after disabling the addons. The whole point of quickFilters is creating and managing filters. not doing handholding while they run. FQ is a bit different as it adds new search terms / actions so it can be actively involved (and filters with such custom parts will not run at all if FQ is not present or disabled)

> 
>     5. Might everything break if there is a Move Message and the folder no longer exists?

the filter will certainly not execute successfully. but `everything` is a thick concept, what do you mean? I would expect other filters to still execute as expected.

> 
>     6. Does this only happen when Move Message is the first action, or second?

no idea - the order of actions within a filter is a mystery and I don't trust it. Thunderbird may prioritize a certain action over another, but it might be nice and tell you about it if it thinks it hast to (at least I have seen similar messages)

> 
>     7. Does it still happen if all Move Message actions are changed to Move Later?

this is a good question, I think move message later was specifically written for time expensive search terms such as reading the whole body with a regex. that part is still unreliable and error prone, it needs thorough reviews. And it is also possible that it needs specific changes for each ESR. Regex and Body and Attachments are reason for me to never monetize FiltaQuilla, I do not want to be responsible for something this undocumented and volatile.
If Tb could add an API that they assume / guarantee to work I would be a much happier camper.

> 
>     8. Might this happen if messages are inbound from the server at the same time that other messages are being manually filtered?

maybe. 

note1: I would add whether this might be more likely on IMAP servers and its settings (such as download message headers only or message size restrictions). I am not sire whether an action within a filter can change existing settings and request a full message download, and I am not certain that this might or might not crash or slow down Thunderbird to a grinding halt. Message synchronisation seems to be risky. I think partly this is owed to the synchronous nature of filter actions. note that if you call an async function from a sync function, the async function gets triggered but returns immediately (practically always before it finished execution) and then the filter may assume it is finished or it may try execute other actions (maybe it will try to move the message?) leading to all sorts of lock ups in an essentially single threaded JavaScript environment.
FILTERS NEED to BE CONVERTED to BECOME ASYNC. It;s my biggest request to Thunderbird to get a chance to make filters stable.

note2: most extension APIs are async, so in order to call async functions during filter actions in a meaningful way (or to use async functions during a search term) the filter interface needs to be converted to async first. So before FILTERS BECOME ASYNC there simply will be no extensions API methods for filters.

> 
>     9. What other pattern might be valid?
> 

Good questions. I hope that the points I made above add some cursory understanding to the architecture, because simple assumptions such as "filter actions are executed one after the other" cannot be verified without a deep look at the underlying methods and code.

**Reply by RealRaven2000** [reply]
> this is a good question, I think move message later was specifically written for time expensive search terms such as reading the whole body with a regex. that part is still unreliable and error prone, it needs thorough reviews. And it is also possible that it needs specific changes for each ESR. Regex and Body and Attachments are reason for me to never monetize FiltaQuilla, I do not want to be responsible for something this undocumented and volatile.

this section sounds a bit harsh and warrants some clarification. the part that is unreliable is streaming the message body, I think there are some new bugs related to this which may be regressions only hitting THunderbrid 115 users. Worst case we may need to roll back some code. Reading the Mime body during filtering is very hard for some reason, I think the message hasn't arrived in the Thunderbird database or isn't accessible through the standard means. I don't know a specific THunderbird developer I can ask about this topic so I am largely in the dark here and my code is hacky when I am trying to read body, I think same would also apply to attachments. as regards documentation, I refer to Thunderbird core code, but I may also be wrong as I haven't fully dived down into the rabbitt hole. 

I am pretty sure the only thing to go by are function header comments and it requires a detailed understanding of all moving parts and different kinds of implementations which are also dependent on different mail protocols. This (the filter service) is all or largely C++, Netscape legacy code, which is hard to read and understand, as an Add-on dev I would really like to stay away from this.

**Reply by RealRaven2000** [reply]
as a joke I would suggest rebranding FiltaQuilla as "the can of worms", congrats for opening it :D :D

**Reply by TonyGravagno** [reply]
> "the can of worms"

Truly LOL here, as before reading this I actually used the same term in my last comment in another thread.  
I believe you and I are running synchronously. 😁

**Reply by TonyGravagno** [reply]
I completely understand and empathize with your notes on topics about the filter service, regressions, body MIME parts, Netscape legacy, async, and desire for a stable API to avoid the rabbit hole(s). So no other comments from me on any of that for now.

**Comment by TonyGravagno** [comment]
I wrote:

> What I've seen: One or more filters can process a mail item, get a successful match which stops further match testing, but then the actions aren't processed, except for Stop Filter Execution. "Train as ..." and "Move Message" are not processed.

And I was wrong about "except for Stop Filter Execution".

Filters will process (`goDoCommand("cmd_applyFilters");` runs) but :  
1) Even when a search term is matched none of the actions are processed.
2) (It therefore follows that) Even when a search term is matched, other filters continue to process because the "Stop Filter Execution" is not executed.

So if we ever talk about "filters not processing" or "filters just stop working" as in the title of this discussion, that is incorrect. More correctly the issue is that **every filter is being processed but all actions stop processing**.

**Reply by RealRaven2000** [reply]
> Filters will process (`goDoCommand("cmd_applyFilters");` runs) but :
> 
>     1. Even when a search term is matched none of the actions are processed.
> 
>     2. (It therefore follows that) Even when a search term is matched, other filters continue to process because the "Stop Filter Execution" is not executed.

my guess: maybe Thunderbird does not set aside / blacklist these "stopped" mails but just keeps testing all new emails with the filter conditions. then it must decide whether a matched mail that it found fitting with another filter had the "stop execution" action applied and drops it from the action.

**Comment by TonyGravagno** [comment]
I believe I've verified that the "actions stop processing" issue can be (or might always be?) triggered during filter processing itself.

I had a folder with two mail items. I ran all filters on the folder (not selected filters or selected mail items). One mail item was fully processed, actions executed, the mail item was moved to another folder, that sequence of filtering ended. During this same run through the folder, the action processing broke, and while the second mail item did match a search term, the actions did not process, and the mail item did not move from that folder.

On Tb restart and a pause to allow normal initialization and cleanup, I re-ran all filters on the folder and the one item processed correctly.

Today I will experiment with copying items to another folder and then running again, to see if the issue is with the folder, the profile, etc. I've tried folder compacting and a couple other things that do not fix the problem.

**Comment by TonyGravagno** [comment]
I am running the Tb/browser debugger / error console. I do not see any related errors. Either no error is being thrown, or I don't have the right debug flags set. Or there could be a silent bug, like some "okToProcess" flag being disabled and the code doing exactly what it is being told to do without throwing an error. Or some idiot could have trapped and silenced an error in a try/catch. We simply don't know about what this blackbox isn't telling us yet.

Can anyone suggest a collection of prefs/flags for verbose tracing?

**Reply by RealRaven2000** [reply]
> Can anyone suggest a collection of prefs/flags for verbose tracing?

I believe there is a command line parameter for Thunderbird itself. And for filtering this may actually log all actions and searches in a separate file. but I cannot remember what the setting is, so I might ask John at the next meeting, if this debug feature still exists.

Maybe have a look at [Thunderbird logging types](https://wiki.mozilla.org/MailNews:Logging#Thunderbird_Logging_Types) in the meantime.

there is also a Thunderbird module name "Filters". I think you need to set the environment variables `MOZ_LOG` and `MOZ_LOG_FILE` to configure Gecko logging for the low level information that happens during filtering. Expect a really big file. The log file is overwritten every time Thunderbird starts new.
